### PR TITLE
Added C# example of graceful shutdown of a AWS Lambda

### PR DESCRIPTION
**Pull request by meijeran**
*Description of changes:*
Added a C# example of graceful shutdown of a AWS Lambda

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by bnusunny — CHANGES_REQUESTED** [review]
Can you move this line into the constructor? This only needs to run once when an instance is created.

**Comment by meijeran** [comment]
@bnusunny I didn't what line you mentioned but I assume it was about the handler registration. I moved it to the constructor.

File: csharp-demo/template.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            Path: /hello
+            Method: get
+
+Outputs:
+  MyDotNetFunctionApi:
+    Description: "API Gateway endpoint URL for Prod stage for MyDotNetFunction"
+    Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
+  MyDotNetFunctionArn:
+    Description: "MyDotNetFunction Lambda Function ARN"
+    Value: !GetAtt MyDotNetFunction.Arn
```
Comment (bnusunny) [inline csharp-demo/template.yaml:34]: `MyDotNetFunction` does not exist. This should be `HelloWorldFunction`.

File: csharp-demo/template.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+Resources:
+  HelloWorldFunction:
+    Type: AWS::Serverless::Function
+    Properties:
+      CodeUri: src/HelloWorldFunction
+      Handler: HelloWorld.Function::FunctionHandler
+      Runtime: dotnet6
+      MemorySize: 512
+      Layers:
+        - arn:aws:lambda:${AWS::Region}:580247275435:layer:LambdaInsightsExtension:14
```
Comment (bnusunny) [inline csharp-demo/template.yaml]: Should add `!Sub` here to use cloudformation variables. 

```yaml
        - !Sub arn:aws:lambda:${AWS::Region}:580247275435:layer:LambdaInsightsExtension:14
 ```

File: csharp-demo/template.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+Globals:
+  Function:
+    Timeout: 20
+
+Resources:
+  HelloWorldFunction:
+    Type: AWS::Serverless::Function
+    Properties:
+      CodeUri: src/HelloWorldFunction
+      Handler: HelloWorld.Function::FunctionHandler
```
Comment (bnusunny) [inline csharp-demo/template.yaml]: Handler is not correct. Should be the following: 

```yaml
      Handler: HelloWorldFunction::HelloWorldFunction.Function::FunctionHandler
```

File: csharp-demo/template.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+Outputs:
+  MyDotNetFunctionApi:
+    Description: "API Gateway endpoint URL for Prod stage for MyDotNetFunction"
+    Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
+  MyDotNetFunctionArn:
+    Description: "MyDotNetFunction Lambda Function ARN"
+    Value: !GetAtt MyDotNetFunction.Arn
+  MyDotNetFunctionIamRole:
+    Description: "Implicit IAM Role created for MyDotNetFunction"
+    Value: !GetAtt MyDotNetFunctionRole.Arn
```
Comment (bnusunny) [inline csharp-demo/template.yaml:37]: Same here. `MyDotNetFunction` does not exist.

File: csharp-demo/README.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+```
+
+Use the following AWS SAM CLI commands to build and deploy this demo.
+
+```bash
+sam build --use-container
+sam deploy --guided 
+```
+
+Take note of the output value of HelloWorldApi. Use curl to invoke the api and trigger the lambda function once.
```
Comment (bnusunny) [inline csharp-demo/README.md]: `HelloWorldApi` should match the Output in Sam template.

File: csharp-demo/README.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+```bash
+curl "replace this with value of HelloWorldApi"
```
Comment (bnusunny) [inline csharp-demo/README.md]: `HelloWorldApi` should match the Output in Sam template.

File: csharp-demo/src/HelloWorldFunction/Function.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    /// <summary>
+    /// A simple function that takes a string and does a ToUpper
+    /// </summary>
+    /// <param name="request"></param>
+    /// <param name="context"></param>
+    /// <returns></returns>
+    public string? FunctionHandler(APIGatewayProxyRequest request, ILambdaContext context)
+    {   
+        return request.Body?.ToUpper();
```
Comment (bnusunny) [inline csharp-demo/src/HelloWorldFunction/Function.cs]: This response does not match API Gateway proxy response format. Could be updated as something like below. 

```csharp
    public APIGatewayProxyResponse FunctionHandler(APIGatewayProxyRequest request, ILambdaContext context)
    {   
        return new APIGatewayProxyResponse{
            StatusCode = 200,
            Body = "Hello world!",
        };
         
    }
```

**Review by bnusunny — CHANGES_REQUESTED** [review]
A few minor issues. Could you update the PR?

**Comment by bnusunny** [comment]
Looks good. Thank you for the contribution!

**Pull request merged**